From 49aa70a75bcc1236ac08cc7818761e70604f7049 Mon Sep 17 00:00:00 2001
From: jenisys <jenisys@users.noreply.github.com>
Date: Mon, 8 Jul 2019 00:38:26 +0200
Subject: [PATCH] gherkin: Adding Rule keyword translation in portuguese and
 spanish to gherkin-languages.json

* Integrate changes based on merged cucumber pull-request #621
* Update "gherkin-languages.json"
* Update "behave/i18n.py" (generated from: gherkin-languages.json)

RELATED-TO: pull #751 (same; using the official way)
---
 CHANGES.rst                        |  1 +
 behave/fixture.py                  |  1 -
 behave/i18n.py                     |  4 +--
 etc/gherkin/gherkin-languages.json |  4 +--
 invoke.yaml                        |  4 +++
 tasks/__init__.py                  |  3 ++
 tasks/develop.py                   | 58 ++++++++++++++++++++++++++++++
 tasks/py.requirements.txt          |  3 ++
 8 files changed, 73 insertions(+), 5 deletions(-)
 create mode 100644 tasks/develop.py

diff --git a/CHANGES.rst b/CHANGES.rst
index 312cbba..15a4ef9 100644
--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -43,6 +43,7 @@ FIXED:
 
 MINOR:
 
+* pull  #751: gherkin: Adding Rule keyword translation in portuguese and spanish to gherkin-languages.json (provided by: dunossauro)
 * pull  #660: Fix minor typos (provided by: rrueth)
 * pull  #655: Use pytest instead of py.test per upstream recommendation (provided by: scop)
 * issue #654: tox.ini: pypi.python.org -> pypi.org (submitted by: pradyunsg)
diff --git a/behave/fixture.py b/behave/fixture.py
index 21093b0..3a9f1bc 100644
--- a/behave/fixture.py
+++ b/behave/fixture.py
@@ -348,7 +348,6 @@ def use_composite_fixture_with(context, fixture_funcs_with_params):
     return composite_fixture
 
 
-
 # -------------------------------------------------------------------------------
 # DECORATORS:
 # -------------------------------------------------------------------------------
diff --git a/behave/i18n.py b/behave/i18n.py
index 721c4c3..2781afe 100644
--- a/behave/i18n.py
+++ b/behave/i18n.py
@@ -331,7 +331,7 @@ languages = \
         'given': ['* ', 'Dado ', 'Dada ', 'Dados ', 'Dadas '],
         'name': 'Spanish',
         'native': 'español',
-        'rule': ['Rule'],
+        'rule': ['Regla'],
         'scenario': ['Ejemplo', 'Escenario'],
         'scenario_outline': ['Esquema del escenario'],
         'then': ['* ', 'Entonces '],
@@ -762,7 +762,7 @@ languages = \
         'given': ['* ', 'Dado ', 'Dada ', 'Dados ', 'Dadas '],
         'name': 'Portuguese',
         'native': 'português',
-        'rule': ['Rule'],
+        'rule': ['Regra'],
         'scenario': ['Exemplo', 'Cenário', 'Cenario'],
         'scenario_outline': ['Esquema do Cenário',
                              'Esquema do Cenario',
diff --git a/etc/gherkin/gherkin-languages.json b/etc/gherkin/gherkin-languages.json
index 913cfac..29cbca1 100644
--- a/etc/gherkin/gherkin-languages.json
+++ b/etc/gherkin/gherkin-languages.json
@@ -1084,7 +1084,7 @@
     "name": "Spanish",
     "native": "español",
     "rule": [
-      "Rule"
+      "Regla"
     ],
     "scenario": [
       "Ejemplo",
@@ -2553,7 +2553,7 @@
     "name": "Portuguese",
     "native": "português",
     "rule": [
-      "Rule"
+      "Regra"
     ],
     "scenario": [
       "Exemplo",
diff --git a/invoke.yaml b/invoke.yaml
index 3e93cfc..d6f141c 100644
--- a/invoke.yaml
+++ b/invoke.yaml
@@ -38,6 +38,10 @@ cleanup:
       - "__WORKDIR__"
       - reports
 
+    extra_files:
+      - "etc/gherkin/gherkin*.json.SAVED"
+      - "etc/gherkin/i18n.py"
+
 cleanup_all:
     extra_directories:
       - .hypothesis
diff --git a/tasks/__init__.py b/tasks/__init__.py
index 969a94a..a572465 100644
--- a/tasks/__init__.py
+++ b/tasks/__init__.py
@@ -39,6 +39,8 @@ from . import _tasklet_cleanup as cleanup
 from . import docs
 from . import test
 from . import release
+from . import develop
+
 
 # -----------------------------------------------------------------------------
 # TASKS:
@@ -56,6 +58,7 @@ namespace.add_collection(Collection.from_module(cleanup), name="cleanup")
 namespace.add_collection(Collection.from_module(docs))
 namespace.add_collection(Collection.from_module(test))
 namespace.add_collection(Collection.from_module(release))
+namespace.add_collection(Collection.from_module(develop))
 cleanup.cleanup_tasks.add_task(cleanup.clean_python)
 
 cleanup.cleanup_tasks.add_task(cleanup.clean_python)
diff --git a/tasks/develop.py b/tasks/develop.py
new file mode 100644
index 0000000..b08df0e
--- /dev/null
+++ b/tasks/develop.py
@@ -0,0 +1,58 @@
+# -*- coding: UTF-8 -*-
+"""
+Development tasks
+"""
+
+from __future__ import absolute_import, print_function
+from invoke import Collection, task
+from invoke.util import cd
+from path import Path
+import requests
+
+# -----------------------------------------------------------------------------
+# CONSTANTS:
+# -----------------------------------------------------------------------------
+GHERKIN_LANGUAGES_URL = "https://raw.githubusercontent.com/cucumber/cucumber/master/gherkin/gherkin-languages.json"
+
+
+# -----------------------------------------------------------------------------
+# TASKS:
+# -----------------------------------------------------------------------------
+@task(name="update_gherkin") # TOO-LONGS: aliases=["update_gherkin_languages"])
+def update_gherkin_languages(ctx):
+    """Update "gherkin-languages.json" file from cucumber-repo."""
+    with cd("etc/gherkin"):
+        # -- BACKUP-FILE:
+        gherkin_languages_file = Path("gherkin-languages.json")
+        gherkin_languages_file.copy("gherkin-languages.json.SAVED")
+
+        print('Downloading "gherkin-languages.json" from github:cucumber ...')
+        download_request = requests.get(GHERKIN_LANGUAGES_URL)
+        gherkin_languages_newfile = Path("gherkin-languages.json.NEW")
+        assert download_request.ok
+        print('Download finished: OK (size={0})'.format(len(download_request.content)))
+        with open(gherkin_languages_newfile, "wb") as f:
+            f.write(download_request.content)
+        gherkin_languages_newfile.rename("gherkin-languages.json")
+
+        print('Generating "i18n.py" ...')
+        ctx.run("./convert_gherkin-languages.py")
+
+
+# -----------------------------------------------------------------------------
+# TASK HELPERS:
+# -----------------------------------------------------------------------------
+def print_packages(packages):
+    print("PACKAGES[%d]:" % len(packages))
+    for package in packages:
+        package_size = package.stat().st_size
+        package_time = package.stat().st_mtime
+        print("  - %s  (size=%s)" % (package, package_size))
+
+
+# -----------------------------------------------------------------------------
+# TASK CONFIGURATION:
+# -----------------------------------------------------------------------------
+namespace = Collection()
+namespace.add_task(update_gherkin_languages)
+namespace.configure({})
diff --git a/tasks/py.requirements.txt b/tasks/py.requirements.txt
index e772d5e..a77d3bc 100644
--- a/tasks/py.requirements.txt
+++ b/tasks/py.requirements.txt
@@ -16,3 +16,6 @@ six >= 1.12.0
 # -- PYTHON2 BACKPORTS:
 pathlib;    python_version <= '3.4'
 backports.shutil_which; python_version <= '3.3'
+
+# -- SECTION: develop
+requests
